### PR TITLE
[feat]: moving averages for LPL learning rule

### PR DESCRIPTION
**Pull request by and-rewsmith**
Implementing the moving averages for the LPL learning rule. 
- Temporal filter
- Spike moving average
- Variance moving average

### CODE REVIEW

File: model/util.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        from reacting too strongly to what might be noise. Conversely, for a
+        system with slow-moving dynamics, you might choose larger values for
+        both tau_rise and tau_fall to avoid reacting to insignificant changes.
+        """
+        self.rise: float = 0
+        self.fall: float = 0
+        self.tau_rise = tau_rise
+        self.tau_fall = tau_fall
+
+    def apply(self, error: float, dt: float = 1) -> float:
```
Comment (mstormbull) [inline model/util.py]: This is looking great. One minor suggestion, nice to hint at units on dt and taurise and taufall if we intend to match units from literature.

File: model/util.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        # Apply the exponential decay to the rise state and add the error
+        self.rise = self.rise * math.exp(-dt / self.tau_rise) + error
+
+        # Apply the exponential decay to the fall state and add the rise state
+        self.fall = self.fall * math.exp(-dt / self.tau_fall) + self.rise
+
+        return self.fall
```
Comment (mstormbull) [inline model/util.py]: LGTM

File: model/tests/test_util.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+
+def test_temporal_filter() -> None:
+    tf = TemporalFilter(tau_rise=1, tau_fall=1)
+
+    # Initial apply since temporal filter is second order
+    tf.apply(error=1)
+
+    # Apply a single error with a small dt should result in small change
+    assert tf.apply(error=1) == pytest.approx(1.7357588823428847)
```
Comment (mstormbull) [inline model/tests/test_util.py:13]: Would this be more intuitive if written in terms of math.exp()?

**Review by mstormbull — APPROVED** [review]
LGTM

File: model/util.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+    def apply(self, spike: float, spike_moving_average: float, dt: float = 1) -> float:
+
+        # Apply the exponential decay to the variance state and add the squared deviation
+        self.variance += (dt / self.tau_var) * \
+            ((spike - spike_moving_average) ** 2 - self.variance)
+
+        return self.variance
+
+    def tracked_value(self) -> float:
```
Comment (and-rewsmith) [inline model/util.py]: @mstormbull We should consider later adding some interface for these. They look similar, but are parameterized differently for the `apply()` fns. Something to keep in mind.

File: model/util.py (hunk 1 above)
Comment (and-rewsmith) [reply model/util.py]: Yeah I can incorporate this into a later diff here where I am touching these files again.
https://github.com/and-rewsmith/LPL-SNN/pull/23

File: model/tests/test_util.py (hunk 3 above)
Comment (and-rewsmith) [reply model/tests/test_util.py:13]: Yes I think that this is a smell that it isn't more intuitive.

The result of `tf.apply()` will be resultant from that iterative update of `tf.rise` and `tf.fall` class vars.  I couldn't think of a way to do this more cleanly. We could pull out the logic from the implementation of `tf.apply()`, but doing it for multiple timesteps of `apply()` would end up with a huge chain of operations. 

So in that case I'd rather have the number. The thought being: looks like it works now, this code probably won't change much once we get the filter ironed out so this number smell will cause minimal disruption.

If you think I am missing something lmk!

**Pull request merged**